### PR TITLE
fix #1814: Avoid infinite recursion when resolving java.lang.Object

### LINKED ISSUE #1814: StackOverflowError in getAllMethodsVisibleToInheritors() (JSS)
Hey,

I just recently upgraded to JavaParser 3.6.19 which broke several of my Unit Tests:
```text
java.lang.StackOverflowError
	at java.base/sun.net.www.protocol.file.Handler.parseURL(Handler.java:67)
	at java.base/java.net.URL.<init>(URL.java:656)
	at java.base/java.net.URL.<init>(URL.java:523)
	at java.base/jdk.internal.loader.URLClassPath$FileLoader.getResource(URLClassPath.java:1127)
	at java.base/jdk.internal.loader.URLClassPath$FileLoader.findResource(URLClassPath.java:1116)
	at java.base/jdk.internal.loader.URLClassPath.findResource(URLClassPath.java:281)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findResourceOnClassPath(BuiltinClassLoader.java:479)
	at java.base/jdk.internal.loader.BuiltinClassLoader.findResource(BuiltinClassLoader.java:303)
	at java.base/java.lang.ClassLoader.getResource(ClassLoader.java:1393)
	at java.base/java.lang.ClassLoader.getResourceAsStream(ClassLoader.java:1729)
	at util.codex.solve.JarSourcesTypeSolver.tryToSolveTypeUncached(JarSourcesTypeSolver.java:28)
	at util.codex.solve.AbstractCodexTypeSolver.tryToSolveType(AbstractCodexTypeSolver.java:85)
	at com.github.javaparser.symbolsolver.resolution.typesolvers.CombinedTypeSolver.tryToSolveType(CombinedTypeSolver.java:62)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.CompilationUnitContext.solveType(CompilationUnitContext.java:136)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.JavaParserTypeDeclarationAdapter.solveType(JavaParserTypeDeclarationAdapter.java:73)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.ClassOrInterfaceDeclarationContext.solveType(ClassOrInterfaceDeclarationContext.java:93)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractMethodLikeDeclarationContext.solveType(AbstractMethodLikeDeclarationContext.java:93)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.convertToUsage(JavaParserFacade.java:526)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.convert(JavaParserFacade.java:586)
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserMethodDeclaration.getReturnType(JavaParserMethodDeclaration.java:71)
	at com.github.javaparser.resolution.MethodUsage.<init>(MethodUsage.java:56)
	at com.github.javaparser.symbolsolver.model.typesystem.ReferenceTypeImpl.getDeclaredMethods(ReferenceTypeImpl.java:150)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.getAllMethodsVisibleToInheritors(ResolvedReferenceType.java:406)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.lambda$getAllMethodsVisibleToInheritors$9(ResolvedReferenceType.java:412)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1378)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.getAllMethodsVisibleToInheritors(ResolvedReferenceType.java:411)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.lambda$getAllMethodsVisibleToInheritors$9(ResolvedReferenceType.java:412)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1378)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.getAllMethodsVisibleToInheritors(ResolvedReferenceType.java:411)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.lambda$getAllMethodsVisibleToInheritors$9(ResolvedReferenceType.java:412)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1378)
	at com.github.javaparser.resolution.types.ResolvedReferenceType.getAllMethodsVisibleToInheritors(ResolvedReferenceType.java:411)
// rest omitted for brevity
```

I believe this is related to https://github.com/javaparser/javaparser/pull/1789: Somehow an infinite recursion occurs when the TypeSolver stumbles upon `java.lang.Object`.

I will try to create a minimal test case as reproduction when I've got time for it.

### PR DESCRIPTION
**Pull request by dhardtke**
This was a surprisingly easy fix.

### CODE REVIEW

File: javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import org.junit.Test;
+
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Dominik Hardtke
+ * @since 01.09.2018
```
Comment (ftomassetti) [inline javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java]: This is how you write dates in German :)
It should be 01/09/2018

File: javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ */
+public class Issue1814 extends AbstractResolutionTest {
+    private JavaParser javaParser;
+
+    @Before
+    public void setup() {
+        // construct a fake java.lang.Object class with only one method (java.lang.Object#equals(java.lang.Object)
+        ClassOrInterfaceDeclaration fakeObject = new ClassOrInterfaceDeclaration();
+        fakeObject.setName(new SimpleName("java.lang.Object"));
```
Comment (ftomassetti) [inline javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java]: I think SimpleName should not contain dots. You may want to set a package declaration instead.

**Review by ftomassetti — APPROVED** [review]
It makes sense to me, I am not sure I would write the tests as it is written but the correction seems correct to me because of the way the name of the class is set. You may want to use some stub or set a package declaration to obtain the fully qualified name you are looking for.

Thank you for your contribution

File: javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java (hunk 2 above)
Comment (dhardtke) [reply javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java]: How would I do that? I didn't find a way to associate a `PackageDeclaration` with my `ClassOrInterfaceDeclaration`

Comment (ftomassetti) [reply javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/Issue1814.java]: You create a CompilationUnit then you set the PackageDeclaration in the CompilationUnit and insert the ClassOrInterfaceDeclaration in the same CompilationUnit. When we want to find the qualified name of a class we do that by looking in which CompilationUnit is contained and reading the associated package declaration

**Comment by ftomassetti** [comment]
Looks good and tests are passing. Thanks!

**Pull request merged**